tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+Tests_SRS_INTERLOCKED_43_036: [ interlocked_and_8 shall return the initial value of *destination. ]
+*/
+TEST_FUNCTION(interlocked_and_8_does_bitwise_and)
+{
+    ///arrange
+    volatile uint8_t destination = 0xF0;
+    uint8_t value = 0x0F;
+
+    ///act
+    uint8_t return_val = (uint8_t)interlocked_and_8((volatile int8_t*)&destination, (int8_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint8_t, 0, destination);
+    ASSERT_ARE_EQUAL(uint8_t, 0xF0, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_006 [ interlocked_compare_exchange shall compare the 32-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.*]
+Tests_SRS_INTERLOCKED_43_037: [ interlocked_compare_exchange shall return the initial value of *destination.]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_equal_case)
+{
+    ///arrange
+    volatile int32_t destination = INT32_MAX;
+    int32_t comperand = INT32_MAX;
+    int32_t exchange = INT32_MIN;
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MIN, destination);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_006 [ interlocked_compare_exchange shall compare the 32-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.*]
+Tests_SRS_INTERLOCKED_43_037: [ interlocked_compare_exchange shall return the initial value of *destination.]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_not_equal_case)
+{
+    ///arrange
+    volatile int32_t destination = INT32_MAX;
+    int32_t comperand = INT32_MIN;
+    int32_t exchange = INT32_MIN;
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, destination);
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_007 [ interlocked_compare_exchange_128 shall compare *destination with *comperand_result. If they are equal, destination[0] is set to exchange_low and destination[1] is set to exchange_high. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_039: [ interlocked_compare_exchange_128 shall store the initial value of *destination in *comperand_result regardless of the result of the comparison.` ]
+Tests_SRS_INTERLOCKED_43_038: [ interlocked_compare_exchange_128 shall return true if *comperand_result equals the original value of *destination.]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_128_equal_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
```

Comment:
>_aligned_malloc [](start = 55, length = 15)

why not volatile int64_t destination[2]?

Reply:
sort of here and everywhere

---
In reply to: [427722625](https://github.com/Azure/azure-c-util/pull/43#discussion_r427722625) [](ancestors = 427722625)

Reply:
I tried creating the array on the stack and got this error: Unhandled exception thrown: read access violation. **destination** was 0xFFFFFFFFFFFFFFFF. occurred

---
In reply to: [427722665](https://github.com/Azure/azure-c-util/pull/43#discussion_r427722665) [](ancestors = 427722665,427722625)